### DISCUSSION TITLE
ISymbol.GetDocumentationCommentXml(...) & <inheritdoc>

### DISCUSSION THREAD

**Discussion by MartyIX** [dotnet/roslyn#50192] [Q&A]
Hi Roslyn team,

https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.isymbol.getdocumentationcommentxml?view=roslyn-dotnet is useful method. As far as I know it does not expand `<inheritdoc cref="..."/>`. Can some Roslyn API help me with expanding of the `<inheritdoc />` or do I need to program it myself?

Thank you

PS: Microsoft Visual Studio 2019 shows in tooltips an expanded form of XML documentations so a reference to that code may be helpful.

EDIT: I have found https://github.com/dotnet/roslyn/blob/73864c784f1b26556dc078a6442527c4bc548f20/src/Workspaces/Core/Portable/Shared/Extensions/ISymbolExtensions.cs#L239 and one can probably use reflection to access the method (but it looks still non-trivial). Unfortunately, it looks it is very hard to get only useful bits from the class.

**Comment by CyrusNajmabadi** [comment]
I don't believe there is any public way to do this. This is effectively an ide feature only currently. @sharwell for thoughts here and if we would want to expose something publicly to help.

**Comment by sharwell** [comment]
The current implementation in `ISymbolExtensions` is not suitable because it only implements the most common parts of the intended full feature. It was only meant to hold us over until the compiler completes the design and implementation for a full language feature. For Roslyn to make this a public API, two things would need to change:

1. The `GetDocumentationCommentXml` method linked above would need new overload(s) to provide the new functionality (would need a new issue filed requesting the compiler feature)
2. The compiler would need to start including documentation files for metadata references consistently during builds (already filed as #23673)

Until that work is complete, the best option is to implement your own version of this in the custom tool, which is the approach taken today by documentation site generators.

**Reply by MartyIX** [reply]
@sharwell Could you possibly file the issue (given that you have good understanding of that)? Or do you think that there is not sufficient demand for this feature?

**Reply by sharwell** [reply]
Here are the existing requests:
https://github.com/dotnet/csharplang/issues/313
https://github.com/dotnet/csharplang/pull/1749

They are slightly out of date, but they didn't seem to be getting traction so it didn't seem useful to keep updating them.

**Reply by MartyIX** [reply]
The number 👍s seems to be high ;-)

**Comment by Danielku15** [comment]
It would be a starting point to have the APIs in Roslyn to resolve `inheritdoc` (and likely `include`) tags but what I think it would be even better if you can also choose that the written XML file during compilation of an assembly, does not include `inheritdoc` anymore but a copy of the original documentation. 

Resolving the right documentation is rather a complex thing due to features like the optional `cref` or `xpath` to be considered correctly. XML documentation files might be used by different toolings like Swashbuckle to display docs in a Swagger UI. To support the right documentation resolving in all those tools (especially considering they do not have Roslyn APIs available) would be counterproductive and prone to errors. 

An extension to the `GenerateDocumentationFile` option (additional separate output file?) where `inheritdoc` tags are eliminated could have direct benefit for all documentation generators like Swashbuckle and Docfx.
